### PR TITLE
Pepsi Ping Pong

### PR DESCRIPTION
**Pull request by okiwi6**
## Introduced Changes

QOL improvement for `pepsi`.
Tries to ping the robot before setting network/downloading logs/uploading/flashing/...
That way error messages are more user friendly.

As a nice side-effect, the ping timeout can now be set whereas previously the timeout was implicitly generated by `rsync` or `ssh`.

Furthermore, a `pepsi` ping subcommand is added which also allows setting timeout duration and retries, just to save users from typing the whole ip of a robot they wanted to ping.

Fixes #

## ToDo / Known Issues

*If this is a WIP describe which problems are to be fixed.*

## Ideas for Next Iterations (Not This PR)

*If there are some improvements that could be done in a next iteration, describe them here.*

## How to Test

* Upload to a robot that is reachable over network, nothing is different
* Try to upload to a robot that is not reachable. A proper error message is generated.

### CODE REVIEW

File: tools/pepsi/src/ping.rs [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+            match nao
+                .is_reachable(
+                    arguments.retries,
+                    Duration::from_secs_f32(arguments.timeout),
+                )
+                .await
+            {
+                true => progress.finish_with_success(format!("{} reachable", nao_address.short())),
+                false => progress
+                    .finish_with_error(Report::msg(format!("{} unreachable", nao_address.short()))),
```
Comment (julianschuler) [inline tools/pepsi/src/ping.rs:43]: I think this section can be done a lot cleaner. `progress.finish_with()` could directly take `Nao::new_with_ping()` (mapped to `()`). The only problem here is the missing arguments. Creating a new constructor with arguments instead and removing `is_reachable()` entirely might be the best solution.

**Review by julianschuler — CHANGES_REQUESTED** [review]
Nice addition!

File: tools/pepsi/src/ping.rs (hunk 1 above)
Comment (okiwi6) [reply tools/pepsi/src/ping.rs:43]: Done

File: crates/nao/src/lib.rs [hunk 2]
```diff
@@ -19,6 +23,30 @@ impl Nao {
         Self { host }
     }
 
+    pub async fn new_with_ping(host: Ipv4Addr) -> Result<Self> {
```
[the remark this answers is not in the document]
Comment (okiwi6) [reply crates/nao/src/lib.rs:26]: Done

File: crates/nao/src/lib.rs [hunk 3]
```diff
@@ -19,6 +23,30 @@ impl Nao {
...
+        Ok(nao)
+    }
+
+    pub async fn is_reachable(&self, retries: usize, timeout: Duration) -> bool {
+        let pinger = async {
+            for _ in 0..retries {
+                match surge_ping::ping(IpAddr::V4(self.host), &[]).await {
+                    Ok(_) => return true,
+                    _ => continue,
+                };
```
[the remark this answers is not in the document]
Comment (okiwi6) [reply crates/nao/src/lib.rs:42]: Done

File: tools/pepsi/src/upload.rs [hunk 4]
```diff
@@ -66,7 +66,8 @@ async fn upload_with_progress(
     arguments: &Arguments,
     progress: &Task,
 ) -> Result<()> {
-    let nao = Nao::new(nao_address.ip);
+    progress.set_message("Pinging Nao");
```
[the remark this answers is not in the document]
Comment (okiwi6) [reply tools/pepsi/src/upload.rs]: Done

File: tools/pepsi/src/ping.rs [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+#[derive(Args)]
+pub struct Arguments {
+    /// Number of ping retries
+    #[arg(long, default_value = "1")]
+    pub retries: usize,
+    /// Time after which ping is aborted
+    #[arg(long, default_value = "2.0")]
+    pub timeout: f32,
+    /// The NAOs to upload to e.g. 20w or 10.1.24.22
```
[the remark this answers is not in the document]
Comment (okiwi6) [reply tools/pepsi/src/ping.rs]: Done

File: tools/pepsi/src/main.rs [hunk 6]
```diff
@@ -166,6 +169,8 @@ enum Command {
     Logs(LogsArguments),
     /// Change player numbers of the NAOs in local configuration
     Playernumber(PlayerNumberArguments),
+    /// Tries to ping the NAO
```
[the remark this answers is not in the document]
Comment (okiwi6) [reply tools/pepsi/src/main.rs]: Done

**Review by julianschuler — APPROVED** [review]

**Pull request merged**